sdk/core/core-rest-pipeline/src/policies/multipartPolicy.ts

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { randomUUID, stringToUint8Array } from "@azure/core-util";
+import { BodyPart, HttpHeaders, PipelineRequest, RequestBodyType } from "../interfaces";
+import { PipelinePolicy } from "../pipeline";
+import { toStream, concatenateStreams } from "../util/stream";
+import { isBlob } from "../util/typeGuards";
+
+function generateBoundary(): string {
+  return `----AzSDKFormBoundary${randomUUID()}`;
+}
+
+function encodeHeaders(headers: HttpHeaders): string {
+  let result = "";
+  for (const [key, value] of headers) {
+    result += `${key}: ${value}\r\n`;
+  }
+  return result;
+}
+
+function getLength(
+  source:
+    | (() => ReadableStream<Uint8Array>)
+    | (() => NodeJS.ReadableStream)
+    | Uint8Array
+    | Blob
+    | ReadableStream
+    | NodeJS.ReadableStream
+): number | undefined {
+  if (source instanceof Uint8Array) {
+    return source.byteLength;
+  } else if (isBlob(source)) {
+    // if was created using createFile then -1 means we have an unknown size
+    return source.size === -1 ? undefined : source.size;
+  } else {
+    return undefined;
+  }
+}
+
+function getTotalLength(
+  sources: (
+    | (() => ReadableStream<Uint8Array>)
+    | (() => NodeJS.ReadableStream)
+    | Uint8Array
+    | Blob
+    | ReadableStream
+    | NodeJS.ReadableStream
+  )[]
+): number | undefined {
+  let total = 0;
+  for (const source of sources) {
+    const partLength = getLength(source);
+    if (partLength === undefined) {
+      return undefined;
+    } else {
+      total += partLength;
+    }
+  }
+  return total;
+}
+
+function buildRequestBody(request: PipelineRequest, parts: BodyPart[], boundary: string): void {
```

Comment:
Could you help me understand why we build this by ourselvers not leveraging the form-data lib?

Reply:
The `form-data` lib had limitations where we weren't able to set the name of the file which was a requirement for OpenAI. We tried using the newer `formdata-node` instead but ran into issues due to that being ESM-only (at the time anyway). Now that we have this policy, we can use it for other multipart content types beyond just form data, so there's an extra win there.